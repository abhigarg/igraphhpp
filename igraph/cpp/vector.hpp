/*
 
 vector.hpp ... Wrapper for igraph vectors 
 
 Copyright (C) 2009  KennyTM~
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 */

/**
 \file vector.hpp
 \brief Wrapper for igraph vectors
 \author KennyTM~
 \date June 12th, 2009
 */

#ifndef IGRAPH_VECTOR_HPP
#define IGRAPH_VECTOR_HPP

#include <igraph/igraph.h>
#include <igraph/cpp/common.hpp>
#include <igraph/cpp/exception.hpp>
#include <cstdarg>
#include <cstdio>
#if XXINTRNL_CXX0X
#include <initializer_list>
#endif

namespace igraph {
	
	XXINTRNL_PREPARE_UNDERLYING_TYPES(BasicVector, vector);	
	/**
	 \class Vector
	 \brief Wrapper for igraph vectors
	 
	 This template class is a wrapper of the igraph vectors type.
	 In igraph there are four kinds of vectors: Vectors of Reals, longs, chars and Booleans.
	 */
	template <typename T>
	class BasicVector {
		XXINTRNL_UNDERLYING_TYPE(BasicVector) _;
		
	public:
		MEMORY_MANAGER_INTERFACE_WITH_TEMPLATE(BasicVector, <T>);
		XXINTRNL_WRAPPER_CONSTRUCTOR_INTERFACE(BasicVector, XXINTRNL_UNDERLYING_TYPE(BasicVector));
		
		/// Copy a C array into a Vector.
		BasicVector(T* const array, const long count) MAY_THROW_EXCEPTION;
		
		/// Create a vector with "count" elements.
		BasicVector(const long count) MAY_THROW_EXCEPTION;
		
#if XXINTRNL_CXX0X
		/// Create a Vector using initializer list (C++0x only.)
		BasicVector(::std::initializer_list<T> elements) MAY_THROW_EXCEPTION;
#endif
		
		/**
		 \brief Create a Vector using content of string.
		 
		 Example:
		 \code
		 Vector v = Vector("42 54 64 75");
		 printf("%lg", v.prod());	// prints 10886400
		 \endcode
		 */
		BasicVector(const char* stringized_elements) MAY_THROW_EXCEPTION;
		
		/// Create a vector sequentially between two values inclusively.
		static RETRIEVE_TEMPORARY_CLASS_WITH_TEMPLATE(BasicVector<T>) seq(const T from, const T to) MAY_THROW_EXCEPTION;
		
		/// Create a vector with "count" elements filled with zero.
		__attribute__((deprecated,warning("BasicVector<T>::zeros is deprecated. Use the constructor BasicVector(x) (for x > 0) or BasicVector<T>::n (for x = 0) instead.")))
		static RETRIEVE_TEMPORARY_CLASS_WITH_TEMPLATE(BasicVector<T>) zeros(const long count = 0) MAY_THROW_EXCEPTION;
		
		/// Create a vector with "count" elements filled with zero.
		static RETRIEVE_TEMPORARY_CLASS_WITH_TEMPLATE(BasicVector<T>) n() MAY_THROW_EXCEPTION;
		
		/// Wrap a C array as a Vector. 
		static RETRIEVE_TEMPORARY_CLASS_WITH_TEMPLATE(BasicVector<T>) view(const T* const array, const long count) throw();
		
		BasicVector<T>& null() throw();
		BasicVector<T>& fill(const T e) throw();
		
		T* ptr() throw() { return VECTOR(_); }
		const T& operator[](const long index) const throw() { return VECTOR(_)[index]; }
		T& operator[](const long index) throw() { return VECTOR(_)[index]; }
		
		T e(const long index) const throw();
		T* e_ptr(const long index) const throw();
		BasicVector<T>& set(const long index, const T value) throw();
		T tail() const throw();
		
		void copy_to(T* store) const throw();
		BasicVector<T>& update(const BasicVector<T>& update_from) MAY_THROW_EXCEPTION;
		BasicVector<T>& append(const BasicVector<T>& append_from) MAY_THROW_EXCEPTION;
		void swap(BasicVector<T>& swap_with) MAY_THROW_EXCEPTION;
		
		BasicVector<T>& swap_elements(const long i, const long j) MAY_THROW_EXCEPTION;
		BasicVector<T>& reverse() MAY_THROW_EXCEPTION;
		
		BasicVector<T>& operator+= (const T k) throw();
		BasicVector<T>& operator-= (const T k) throw();
		BasicVector<T>& operator*= (const T k) throw();
		BasicVector<T>& operator/= (const T k) throw();
		
		BasicVector<T>& operator+= (const BasicVector<T>& k) MAY_THROW_EXCEPTION;
		BasicVector<T>& operator-= (const BasicVector<T>& k) MAY_THROW_EXCEPTION;
		BasicVector<T>& operator*= (const BasicVector<T>& k) MAY_THROW_EXCEPTION;
		BasicVector<T>& operator/= (const BasicVector<T>& k) MAY_THROW_EXCEPTION;
		
		T min() const throw();
		T max() const throw();
		long which_min() const throw();
		long which_max() const throw();
		void minmax(T& minStore, T& maxStore) const MAY_THROW_EXCEPTION;
		void which_minmax(long& minStore, long& maxStore) const MAY_THROW_EXCEPTION;
		
		bool empty() const throw();
		long size() const throw();
		
		T sum() const throw();
		T prod() const throw();
		bool isnull() const throw();
		bool isininterval(T low, T high) const throw();
		bool any_smaller(T upper_limit) const throw();
		bool operator== (const BasicVector<T>& other) const throw();
		bool operator!= (const BasicVector<T>& other) const throw();
		T maxdifference(const BasicVector<T>& other) const throw();
		
		bool contains(const T e) const throw();
		bool search(const T what, const long from = 0) const throw();
		bool search(const T what, const long from, long& pos) const throw();
		bool binsearch(const T what) const throw();
		bool binsearch(const T what, long& pos) const throw();
		
		BasicVector<T>& clear() throw();
		BasicVector<T>& reserve(const long new_size) MAY_THROW_EXCEPTION;
		BasicVector<T>& resize(const long new_size) MAY_THROW_EXCEPTION;
		BasicVector<T>& push_back(const T e) MAY_THROW_EXCEPTION;
		T pop_back() throw();
		BasicVector<T>& insert(const long pos, const T e) MAY_THROW_EXCEPTION;
		BasicVector<T>& remove(const long pos) throw();
		BasicVector<T>& remove_section(const long from, const long to) throw();
		
		BasicVector<T>& sort() throw();
				
		// STL support
		typedef T value_type;
		typedef value_type* pointer;
		typedef value_type& reference;
		typedef const value_type& const_reference;
		typedef unsigned size_type;
		typedef int difference_type;
		typedef value_type* iterator;
		typedef const value_type* const_iterator;
		
		iterator begin() throw() { return _.stor_begin; }
		iterator end() throw() { return _.end; }
		const_iterator begin() const throw() { return _.stor_begin; }
		const_iterator end() const throw() { return _.end; }
		size_type capacity() const throw() { return _.stor_end - _.stor_begin; }
		reference front() throw() { return *_.stor_begin; }
		const_reference front() const throw() { return *_.stor_begin; }
		reference back() throw() { return *(_.end-1); }
		const_reference back() const throw() { return *(_.end-1); }
		
		/// Print content of the Vector.
		void print(std::FILE* f = stdout) const throw() {
			bool is_first = true;
			for (const_iterator cit = begin(); cit != end(); ++ cit) {
				if (is_first)
					is_first = false;
				else
					::std::fprintf(f, " ");
				XXINTRNL_fprintf(f, *cit);
			}
			::std::fprintf(f, "\n");
		}
		
		/// Convert from std::vector.
		template<typename InputIterator>
		BasicVector(const InputIterator from, const InputIterator to) MAY_THROW_EXCEPTION;
		
		friend class VertexSelector;
		friend class VertexIterator;
		friend class EdgeSelector;
		friend class EdgeIterator;
		friend class Graph;
		friend class GraphWriter;
		friend class GraphReader;
	};
	MEMORY_MANAGER_INTERFACE_EX_WITH_TEMPLATE(template<typename T>, BasicVector, <T>);
	
	typedef BasicVector<Real> Vector;
	typedef BasicVector<Boolean> BoolVector;
	typedef Vector VertexVector;
	typedef Vector EdgeVector;
}
										   
#endif