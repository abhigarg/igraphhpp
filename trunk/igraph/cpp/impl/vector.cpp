/*

vector.cpp ... Wrapper implementation of igraph vector.

Copyright (C) 2009  KennyTM~

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef IGRAPH_VECTOR_CPP
#define IGRAPH_VECTOR_CPP

#include <igraph/cpp/vector.hpp>
#include <cstring>

namespace igraph {
	
	MEMORY_MANAGER_IMPLEMENTATION_WITH_TEMPLATE(template<typename T>, BasicVector, <T>);
	
#define BASE Real
#define FUNC(x) igraph_vector_##x
#define TYPE FUNC(t)
#include <igraph/cpp/impl/vector.4.cpp>
#undef TYPE
#undef FUNC
#undef BASE

#define BASE long
#define FUNC(x) igraph_vector_long_##x
#define TYPE FUNC(t)
#include <igraph/cpp/impl/vector.4.cpp>
#undef TYPE
#undef FUNC
#undef BASE
	
#define BASE char
#define FUNC(x) igraph_vector_char_##x
#define TYPE FUNC(t)
#include <igraph/cpp/impl/vector.4.cpp>
#undef TYPE
#undef FUNC
#undef BASE
	
#define BASE igraph_bool_t
#define FUNC(x) igraph_vector_bool_##x
#define TYPE FUNC(t)
#include <igraph/cpp/impl/vector.4.cpp>
#undef TYPE
#undef FUNC
#undef BASE	
			
}

#endif