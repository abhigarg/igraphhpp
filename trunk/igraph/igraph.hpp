/*

igraph.hpp ... C++ wrapper for igraph

Copyright (C) 2009  KennyTM~

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef IGRAPH_HPP
#define IGRAPH_HPP


#include <igraph/cpp/common.hpp>
#include <igraph/cpp/exception.hpp>

#include <igraph/cpp/vector.hpp>
#include <igraph/cpp/pointervector.hpp>

#include <igraph/cpp/graph.hpp>
#include <igraph/cpp/graphio.hpp>

#include <igraph/cpp/vertexselector.hpp>
#include <igraph/cpp/vertexiterator.hpp>
#include <igraph/cpp/edgeselector.hpp>
#include <igraph/cpp/edgeiterator.hpp>

//------------------------------------------------------------------------------

#include <igraph/cpp/impl/vector.cpp>
#include <igraph/cpp/impl/pointervector.cpp>

#include <igraph/cpp/impl/vertexselector.cpp>
#include <igraph/cpp/impl/edgeselector.cpp>

#include <igraph/cpp/impl/graphio.cpp>
#include <igraph/cpp/impl/graph.cpp>

#include <igraph/cpp/impl/iterators.cpp>

#endif