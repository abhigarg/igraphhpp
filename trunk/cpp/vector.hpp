/*
 
 vector.hpp ... Wrapper for igraph vectors 
 
 Copyright (C) 2009  KennyTM~
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 */

/**
 \file vector.hpp
 \brief Wrapper for igraph vectors
 \author KennyTM~
 \date June 12th, 2009
 */

#ifndef IGRAPH_VECTOR_HPP
#define IGRAPH_VECTOR_HPP

#include <igraph/igraph.h>
#include <igraph/cpp/common.hpp>
#include <igraph/cpp/exception.hpp>
#include <igraph/cpp/tempobj.hpp>
#include <cstdarg>

namespace igraph {
	/**
	 \class Vector
	 \brief Wrapper for igraph vectors
	 
	 This class is a wrapper of the igraph_vector_t type, which is an array of Real.
	 */
	class Vector {
	private:
		igraph_vector_t _;
		Vector() throw() {};
		
	public:
		MEMORY_MANAGER_INTERFACE(Vector);
		
		/// Create a vector with "count" elements filled with zero.
		Vector(const long count) throw(Exception) {
			TRY(igraph_vector_init(&_, count));
		}		
		
		/// Copy a C array into a Vector.
		Vector(Real* const array, const long count) throw(Exception) {
			TRY(igraph_vector_init_copy(&_, array, count));
		}
		
		/// Create a vector sequentially between two values inclusively.
		static Vector seq(const Real from, const Real to) throw(Exception) {
			Vector v;
			TRY(igraph_vector_init_seq(&v._, from, to));
			return v;
		}
				
		/// Wrap a C array as a Vector. 
		static Vector view(const Real* const array, const long count) throw() {
			Vector v;
			v.mm_dont_dealloc = true;
			igraph_vector_view(&v._, array, count);
			return v;
		}
		
		/// var_arg method to construct the vector
		Vector(const long count, ...) throw(Exception) {
#if __GNUC__ >= 3
			Real arr[count];
#else
			Real* arr = new Real[count];
#endif
			std::va_list va;
			va_start(va, count);
			for (long i = 0; i < count; ++ i)
				arr[i] = va_arg(va, Real);
			va_end(va);
			
			TRY(igraph_vector_init_copy(&_, arr, count));
			
#if !(__GNUC__ >= 3)
			// FIXME: May cause memory leak when an exception is thrown.
			delete[] arr;
#endif
		}
		
		/// Wrap or copy an igraph_vector_t to a Vector.
		Vector(const igraph_vector_t& orig, const OwnershipTransfer transfer = OwnershipTransferKeepOriginal) throw(Exception) : COMMON_INIT_WITH(transfer) {
			if (transfer == OwnershipTransferCopy) {
				TRY(igraph_vector_copy(&_, &orig));
			} else
				_ = orig;
		}
		
		
		void null() throw() { igraph_vector_null(&_); }
		void fill(const Real e) throw() { igraph_vector_fill(&_, e); }
		
		Real* ptr() throw() { return VECTOR(_); }
		const Real& operator[](const long index) const throw() { return VECTOR(_)[index]; }
		Real& operator[](const long index) throw() { return VECTOR(_)[index]; }
		
		Real e(const long index) const throw() { return igraph_vector_e(&_, index); }
		Real* e_ptr(const long index) const throw() { return igraph_vector_e_ptr(&_, index); }
		void set(const long index, const Real value) throw() { igraph_vector_set(&_, index, value); }
		Real tail() const throw() { return igraph_vector_tail(&_); }
		
		void copy_to(Real* store) const throw() { igraph_vector_copy_to(&_, store); }
		void update(const Vector& update_from) throw(Exception) { TRY(igraph_vector_update(&_, &update_from._)); }
		void append(const Vector& append_from) throw(Exception) { TRY(igraph_vector_append(&_, &append_from._)); }
		void swap(Vector& swap_with) throw(Exception) { TRY(igraph_vector_swap(&_, &swap_with._)); }
		
		void swap_elements(const long i, const long j) throw(Exception) { TRY(igraph_vector_swap_elements(&_, i, j)); }
		void reverse() throw(Exception) { TRY(igraph_vector_reverse(&_)); }
		
		Vector& operator+= (const Real k) throw() { igraph_vector_add_constant(&_, k); return *this; }
		Vector& operator-= (const Real k) throw() { igraph_vector_add_constant(&_, -k); return *this; }
		Vector& operator*= (const Real k) throw() { igraph_vector_scale(&_, k); return *this; }
		Vector& operator/= (const Real k) throw() { igraph_vector_scale(&_, 1/k); return *this; }
		
		Vector& operator+= (const Vector& k) throw(Exception) { TRY(igraph_vector_add(&_, &k._)); return *this; }
		Vector& operator-= (const Vector& k) throw(Exception) { TRY(igraph_vector_sub(&_, &k._)); return *this; }
		Vector& operator*= (const Vector& k) throw(Exception) { TRY(igraph_vector_mul(&_, &k._)); return *this; }
		Vector& operator/= (const Vector& k) throw(Exception) { TRY(igraph_vector_div(&_, &k._)); return *this; }
		
		Real min() const throw() { return igraph_vector_min(&_); }
		Real max() const throw() { return igraph_vector_max(&_); }
		long which_min() const throw() { return igraph_vector_which_min(&_); }
		long which_max() const throw() { return igraph_vector_which_max(&_); }
		void minmax(Real& minStore, Real& maxStore) const throw(Exception) { TRY(igraph_vector_minmax(&_, &minStore, &maxStore)); }
		void which_minmax(long& minStore, long& maxStore) const throw(Exception) { TRY(igraph_vector_which_minmax(&_, &minStore, &maxStore)); }
		
		bool empty() const throw() { return igraph_vector_empty(&_); }
		long size() const throw() { return igraph_vector_size(&_); }
		
		Real sum() const throw() { return igraph_vector_sum(&_); }
		Real prod() const throw() { return igraph_vector_prod(&_); }
		bool isininterval(Real low, Real high) const throw() { return igraph_vector_isininterval(&_, low, high); }
		bool operator== (const Vector& other) const throw() { return igraph_vector_is_equal(&_, &other._); }
		bool operator!= (const Vector& other) const throw() { return !(*this == other); }
		Real maxdifference(const Vector& other) const throw() { return igraph_vector_maxdifference(&_, &other._); }
		
		bool contains(const Real e) const throw() { return igraph_vector_contains(&_, e); }
		bool search(const Real what, const long from = 0) const throw() { return igraph_vector_search(&_, from, what, NULL); }
		bool search(const Real what, const long from, long& pos) const throw() { return igraph_vector_search(&_, from, what, &pos); }
		bool binsearch(const Real what) const throw() { return igraph_vector_binsearch2(&_, what); }
		bool binsearch(const Real what, long& pos) const throw() { return igraph_vector_binsearch(&_, what, &pos); }
		
		void clear() throw() { igraph_vector_clear(&_); }
		void reserve(const long new_size) throw(Exception) { TRY(igraph_vector_reserve(&_, new_size)); }
		void resize(const long new_size) throw(Exception) { TRY(igraph_vector_resize(&_, new_size)); }
		void push_back(const Real e) throw(Exception) { TRY(igraph_vector_push_back(&_, e)); }
		Real pop_back() throw() { return igraph_vector_pop_back(&_); }
		void insert(const long pos, const Real e) throw(Exception) { TRY(igraph_vector_insert(&_, pos, e)); }
		void remove(const long pos) throw() { igraph_vector_remove(&_, pos); }
		void remove_section(const long from, const long to) throw() { igraph_vector_remove_section(&_, from, to); }
		
		void sort() throw() { igraph_vector_sort(&_); }
		
		// STL support
		typedef Real value_type;
		typedef value_type* pointer;
		typedef value_type& reference;
		typedef const value_type& const_reference;
		typedef unsigned size_type;
		typedef int difference_type;
		typedef value_type* iterator;
		typedef const value_type* const_iterator;
		
		iterator begin() { return _.stor_begin; }
		iterator end() { return _.end; }
		const_iterator begin() const { return _.stor_begin; }
		const_iterator end() const { return _.end; }
		size_type capacity() const { return _.stor_end - _.stor_begin; }
		reference front() { return *_.stor_begin; }
		const_reference front() const { *_.stor_begin; }
		reference back() { return *(_.end-1); }
		const_reference back() const { return *(_.end-1); }
		
		/// Print content of the Vector.
		void print() const throw() {
			for (const_iterator cit = begin(); cit != end(); ++ cit)
				printf("%lg, ", *cit);
			printf("\n");
		}
		
		/// Convert from std::vector.
		template<typename InputIterator>
		Vector(const InputIterator from, const InputIterator to) throw(Exception) {
			if (from != to) {
				TRY(igraph_vector_init(&_, 1));
				for (const InputIterator cit = from; cit != to; ++ cit)
					TRY(igraph_vector_push_back(&_, *cit));
			} else
				mm_dont_dealloc = true;
		}
		
		friend class VertexSelector;
		friend class VertexIterator;
	};

	MEMORY_MANAGER_IMPLEMENTATION_ATTR(inline, Vector);
	
	IMPLEMENT_COPY_METHOD(Vector) {
		igraph_vector_copy(&_, &other._);
	}
	IMPLEMENT_MOVE_METHOD(Vector) {
		_ = std::move(other._);
	}
	IMPLEMENT_DEALLOC_METHOD(Vector) {
		igraph_vector_destroy(&_);
	}

	IMMEDIATE_OPERATOR_IMPLEMENTATION_COMMUTATIVE_ATTR(static inline, Vector, +, Real);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_RHS_ATTR(static inline, Vector, -, Real);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_COMMUTATIVE_ATTR(static inline, Vector, *, Real);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_RHS_ATTR(static inline, Vector, /, Real);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_ATTR(static inline, Vector, +);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_ATTR(static inline, Vector, -);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_ATTR(static inline, Vector, *);
	IMMEDIATE_OPERATOR_IMPLEMENTATION_ATTR(static inline, Vector, /);
}
										   
#endif