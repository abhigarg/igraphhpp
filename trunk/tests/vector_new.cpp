#undef NDEBUG

#include <cassert>
#include <igraph/igraph.hpp>
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;
using namespace igraph;

int main () {
	Vector v = Vector("4., 5., 6., 9., 8., 7.");
		
	v.print();
	v.print(" .... ");
	v.print(" @@ ", stdout);
	
	assert(v == v);
	assert(v == Vector(" 4., 5., 6., 9., 8., 7."));
	assert(v != Vector(" 3., 4., 5."));
	assert(v.size() == 6);
	assert(v.sum() == 4 + 5 + 6 + 9 + 8 + 7);
	assert(v.prod() == 4 * 5 * 6 * 9 * 8 * 7);
	assert(v.max() == 9);
	assert(v.min() == 4);
	assert(v.which_max() == 3);
	assert(v.which_min() == 0);
	Real v_max, v_min;
	long which_max, which_min;
	v.minmax(v_min, v_max);
	assert(v_min == 4 && v_max == 9);
	v.which_minmax(which_min, which_max);
	assert(which_min == 0 && which_max == 3);
	v += 1;
	assert(v == Vector("5., 6., 7., 10., 9., 8."));
	v *= 2;
	assert(v == Vector("10., 12., 14., 20., 18., 16."));
	v.sort();
	assert(v == Vector("10., 12., 14., 16., 18., 20."));
	
	assert(v.pop_back() == 20.);
	assert(v == Vector(" 10., 12., 14., 16., 18."));
	v.push_back(12.);
	assert(v == Vector(" 10., 12., 14., 16., 18., 12."));
	
	{
		Vector x = v * (3 + v + 2) * v;
		assert(v == Vector("10., 12., 14., 16., 18., 12."));
		assert(x == Vector("1500., 2448., 3724., 5376., 7452., 2448."));
	}
	assert(v == Vector("10., 12., 14., 16., 18., 12."));
	
	Real temp[] = {2, 3, 5, 7, 11};
	{
		Vector w = Vector::view(temp, 5);
		assert(w == Vector("2., 3., 5., 7., 11."));
	}
	assert(temp[0] == 2 && temp[1] == 3 && temp[2] == 5 && temp[3] == 7 && temp[4] == 11);
	
	v.append(Vector("6 7"));
	assert(v == Vector(" 10., 12., 14., 16., 18., 12., 6., 7."));
	assert(v.contains(16.) && v.contains(6.));
	assert(!v.contains(2.));
	long pos = -1;
	assert(v.search(18., 3, pos));
	assert(pos == 4);
	assert(!v.search(18., 5, pos));
	assert(!v.search(9., 0));
	assert(!v.search(10., 1));
	v.sort();
	assert(v.binsearch(16., pos));
	assert(pos == 6);
	assert(!v.binsearch(4.));
	
	v.null();
	assert(v == Vector(" 0., 0., 0., 0., 0., 0., 0., 0."));
	v.fill(12.);
	assert(v == Vector(" 12., 12., 12., 12., 12., 12., 12., 12."));
	
	Vector x = Vector::seq(4, 10);
	assert(x == Vector(" 4., 5., 6., 7., 8., 9., 10."));
	assert(x[2] == 6.);
	x[2] = 13.;
	assert(x[2] == 13.);
	assert(x[3] == 7.);
	assert(x == Vector(" 4., 5., 13., 7., 8., 9., 10."));
	assert(x.e(2) == 13.);
	x.set(3, 14.);
	assert(x == Vector(" 4., 5., 13., 14., 8., 9., 10."));
	assert(*x.e_ptr(3) == 14.);
	*x.e_ptr(4) = 15.;
	assert(x == Vector(" 4., 5., 13., 14., 15., 9., 10."));
	assert(x.tail() == 10.);
	
	Real c[7];
	x.copy_to(c);
	assert(c[0] == 4. && c[1] == 5. && c[2] == 13. && c[3] == 14. && c[4] == 15. && c[5] == 9. && c[6] == 10.);
	x.resize(4);
	assert(x == Vector("4., 5., 13., 14."));
	Vector y = x;
	x.reverse();
	assert(x == Vector(" 14., 13., 5., 4."));
	assert(y == Vector("4., 5., 13., 14."));
	x.swap(y);
	assert(y == Vector("14., 13., 5., 4."));
	assert(x == Vector("4., 5., 13., 14."));
	y.remove(2);
	assert(y == Vector("14., 13., 4."));
	y.swap_elements(0, 1);
	assert(y == Vector("13., 14., 4."));
	x.resize(2);
	assert(x == Vector("4., 5."));
	y.insert(2, 64.);
	y.insert(1, 89.);
	assert(y == Vector("13., 89., 14., 64., 4."));
	{
		Vector u = x;
		assert(x == u);
		u.update(y);
		assert(u == y);
		assert(u == Vector("13., 89., 14., 64., 4."));
	}
	assert(x == Vector("4 5"));
	assert(y == Vector("13., 89., 14., 64., 4."));
	y.remove_section(2, 4);
	assert(y == Vector("13., 89., 4."));
	y.clear();
	assert(y.size() == 0);
	
	{
		igraph_vector_t v6;
		igraph_vector_init(&v6, 7);
		Vector v7(&v6);
		Vector v8=v7;
	}
	
	std::vector<int> sv;
	sv.clear();
	sv.push_back(4);
	sv.push_back(14);
	sv.push_back(5);
	sv.push_back(7);
	
	Vector r = Vector(sv.begin(), sv.end());
	assert(r == Vector("4 14 5 7"));
	sv[1] = 22;
	assert(sv.size() == 4 && sv[0] == 4 && sv[1] == 22 && sv[2] == 5 && sv[3] == 7);
	assert(r == Vector("4 14 5 7"));
	sv = std::vector<int>(r.begin(), r.end());
	assert(sv.size() == 4 && sv[0] == 4 && sv[1] == 14 && sv[2] == 5 && sv[3] == 7);
	r[1] = 22;
	assert(sv.size() == 4 && sv[0] == 4 && sv[1] == 14 && sv[2] == 5 && sv[3] == 7);
	assert(r == Vector("4 22 5 7"));
	std::sort(r.begin(), r.end());
	assert(r == Vector("4 5 7 22"));
	assert(std::inner_product(sv.begin(), sv.end(), r.begin(), 0) == 4*4+14*5+5*7+7*22);
	
	printf("vector.hpp is correct.\n");

	return 0;
}
