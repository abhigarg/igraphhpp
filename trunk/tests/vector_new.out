4 5 6 9 8 7
4 .... 5 .... 6 .... 9 .... 8 .... 7
4 @@ 5 @@ 6 @@ 9 @@ 8 @@ 7
vector.hpp is correct.
